kubernetes: virtual machines - add Vm Detail page

suomiy opened the pull request:
First stage of #8948

- adds Vm Detail page and navigation between vms and vm
- this PR still uses old terminology (vm/ovm), but it will be changed in the next stages

I also created DetailPage component for easier creation of pages like this (will be used at least for Vm/Vm Instance). Not sure if I should split the component to another PR?


Todo: 
- [x] polish ui
- [x] fix tests (mainly ids)
- [x] add tests - basic use case + navigation

suomiy commented:
#### Screenshots:

##### Navigation

- Click on vm in vms list

![screenshot from 2018-06-21 16-21-58](https://user-images.githubusercontent.com/3648838/41725881-382f4d66-7571-11e8-964d-741347a9c240.png)

##### Vm Detail

![screenshot from 2018-06-21 15-52-46](https://user-images.githubusercontent.com/3648838/41725928-4e16c686-7571-11e8-9569-a76d3f87f418.png)

##### 404 page

![not-exists](https://user-images.githubusercontent.com/3648838/41726014-808e06c4-7571-11e8-800b-9d8bf8b0c207.png)

suomiy commented:
Few notes about the design

![screenshot-localhost-9090-2018 06 21-16-30-17](https://user-images.githubusercontent.com/3648838/41726400-571b1a7e-7572-11e8-87c6-831a62878cc3.png)

1. I am showing the icon of vm (will be probably the same for vm/vm instance) and "namespace:name". Not sure if I should put the namespace somewhere else
2. This is the link to go back to the vm list. It is consistent with kubernetes plugin. We could also show a path there
3. There will be Vm / Vm Instance depending on the type here (old terminology is Offline Vm / Vm)
4. I also put state there. Please let me know if I should put it somewhere else.

@garrett @andreasn @beanh66

suomiy commented:
\+ the vm details are accessible at `/kubernetes#/vms/$VM_UID`

mykaul commented:
VM, not Vm.
VMs, not Vms.

I'm also unsure about the title 'Vm' instead of just 'Details' ?

beanh66 commented:
@suomiy For #2 in your list, can we show that as breadcrumbs instead?

suomiy commented:
> VM, not Vm.
> VMs, not Vms.

thanks

> I'm also unsure about the title 'Vm' instead of just 'Details' ?

Could be 'Details' but we have to indicate the difference between 'VM' and 'VM Instance' somehow, because they will have similar pages.

> For #2 in your list, can we show that as breadcrumbs instead?

I can add breadcrumbs. But all of the other pages in Kubernetes plugin are done this way. I am not sure if I should keep it for consistency simpler like this.

There is also another problem:

- breadcrumbs for `VM Instances > Vm Instance` will not have VM Instances page because it will redirect directly into the vms. 
- but it will have it's own path `/kubernetes#/vminstances/$VM_INSTANCE_UID`

andreasn commented:
> I can add breadcrumbs. But all of the other pages in Kubernetes plugin are done this way. I am not sure if I should keep it for consistency simpler like this.

I agree with this. It's important to keep consistency across pages. We can revisit the current pattern later if it performs badly in usability tests, and if so, we can change all the navigation across all pages in the kubernetes plugin to work like that. But for now, I would prefer if the current behaviour is kept.

beanh66 commented:
@suomiy @andreasn Got it, thanks for the explanation! That makes sense to me, keeping consistent behavior in the product. 👍

suomiy commented:
fixed Vm -> 'VM'

mareklibra commented on pkg/kubernetes/scripts/virtual-machines/components/VmDetail.jsx:
```diff
@@ -0,0 +1,103 @@
+            <div>
+                <DetailPage>
+                    {header}
+                    <DetailPageRow title={cockpit.format(_("Vm with uid $0 does not exist."), vmUid)} idPrefix={'vm-not-found'} />
```
nit: `uid` should be in CAPITAL

mareklibra replied:
fine for a follow-up

mareklibra commented on pkg/kubernetes/scripts/virtual-machines/components/VmDisksTabKubevirt.jsx:
```diff
@@ -66,7 +66,7 @@ const prepareDiskData = (disk, vm, pvs, idPrefix) => {
         bus = _("iSCSI");
         onNavigate = () => cockpit.jump(`/kubernetes#/volumes/${pv.metadata.name}`);
-    } else if (disk.disk.bus) {
+    } else if (disk.disk && disk.disk) {
```
Is this what was intended?

mareklibra commented on pkg/kubernetes/scripts/virtual-machines/entry-points/virtual-machine.jsx:
```diff
@@ -0,0 +1,80 @@
+ * @param {kubeMethods} kubeMethods
+ * @param {KubeRequest} KubeRequest
+ */
+function init($scope, $routeParams, kubeLoader, kubeSelect, kubeMethods, KubeRequest) {
```
I think we can share can between these two entry points even more

suomiy replied:
What do you suggest? Unifying the id of the root element and passing the component to render? I do not see other improvements otherwise.

mareklibra commented:
Important to note: this should work for kubevirt `>= 0.6.0`
Oncoming `0.7.0` (so far just pre-release `alpha.3` is out) will change terminology in k8s objects, means  code introduced by this PR will need to be adjusted.

I've tested this PR against `kubevirt 0.6.0`.

Might be my misunderstanding, but
`oc get offlinevirtualmachine.kubevirt.io` shows 2 OVMs but they are not listed within `/kubernetes#/vms` view . Just VM Instances are listed there.

I'm not quite sure if recent 2 entrypoints are good solution. If the user navigates between list and detail, redux store is reinitialized so loosing its state. There's so far no issue with that but I can imagine use-cases where this might be serious issue.

suomiy commented:
> Might be my misunderstanding, but
oc get offlinevirtualmachine.kubevirt.io shows 2 OVMs but they are not listed within /kubernetes#/vms view . Just VM Instances are listed there.

yes this PR is concerned only with Vm Instances (AKA old vms). 

I'm not quite sure if recent 2 entrypoints are good solesution. If the user navigates between list and detail, redux store is reinitialized so loosing its state. There's so far no issue with that but I can imagine use-cases where this might be serious issue.

Yes, but we also want the possibility to just navigate to single vm.

suomiy commented:
updated my last comment and sorry for the misclick

suomiy commented:
- changed navigation from `"/kubernetes#/vms/$VM_UID"` to `"/kubernetes#/vms/$VM_NAMESPACE/$VM_NAME"` to match other kubernetes pages' behaviour. Also it behaves better when restarting vm (vm can have different UID) . - using old terminology

- added tests 

@mareklibra please can you rereview?

suomiy commented:
It seems I am not able to re-trigger the tests with `bots/tests-trigger`. Tests are triggered successfully when I push the changes.
```
./bots/tests-trigger 9466

verify/ubuntu-1604: triggering on pull request 9466
statuses/d76ee4e842ff4ca840f07c6c88273c5d68ded607
{"message":"Not Found","documentation_url":"https://developer.github.com/v3/repos/statuses/#create-a-status"}
Traceback (most recent call last):
  File "./bots/tests-trigger", line 74, in <module>
    sys.exit(main())
  File "./bots/tests-trigger", line 71, in main
    return trigger_pull(api, opts)
  File "./bots/tests-trigger", line 55, in trigger_pull
    api.post("statuses/" + revision, changes)
  File "/home/ansy/Projects/IdeaProjects/cockpit/bots/task/github.py", line 201, in post
    raise RuntimeError("GitHub API problem: {0}".format(response['reason'] or status))
RuntimeError: GitHub API problem: Not Found
``` 
Does somebody have a clue  what might be a problem?

suomiy commented:
i also have refreshed token in ~/.config/github-token

@martinpitt @stefwalter

martinpitt commented:
@suomiy : You are in the [contributors team](https://github.com/orgs/cockpit-project/teams/contributors/members), so I'm afraid I don't have an immediate idea what's wrong.

mareklibra commented:
This should work against kubevirt 0.6.0 - so let's aim at merging this and change later once #9638 lands with `>=0.7.1`

suomiy commented:
@mareklibra fixed

martinpitt commented:
[testKubevirtVmInstance](https://fedorapeople.org/groups/cockpit/logs/pull-9466-20180730-070808-c5ea3bc3-verify-fedora-28/log.html#219) fails, setting needswork. (Let's please not introduce even more flakes, the kubevirt tests are already bad enough for landing PRs.)

suomiy commented:
> testKubevirtVmInstance fails, setting needswork. (Let's please not introduce even more flakes, the kubevirt tests are already bad enough for landing PRs.)

@martinpitt this is actually the same flake as in `testKubevirtMachinesList` because the detail should show the same UI as the list, thus it is reusing the same code and we get the same flake. I can look at the tests and see if I can fix the flakes, but I am not sure if I will be successful, because the tests usually pass on my machine.

martinpitt commented:
@suomiy : OK; let's retry the tests instead, the new test should at least be able to succeed.

For reproducing, try to run the tests with `-j4` or so, to slow them down :-)

suomiy commented:
I tried that (100% usage CPU and 90% RAM) but still no luck. Also got few of these: https://github.com/bcbcarl/react-c3js/issues/22 ,which I hope are not a cause of some of these failures. 

Current test runs seems to have passed fine for `testKubevirtVmInstance `
